### ISSUE TITLE
Migration lints for 2021 prelude

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#84594]
The 2021 Edition plans to have a new prelude, which will include a number of new traits. This means that new methods will be in scope which could introduce ambiguities into some code bases if they are using methods with the same name (but from different traits). We need a migration lint that detects such cases and suggests a rewrite into fully qualified form: e.g., `x.foo()` would become something like `Trait::foo(&x)`.

**Comment by nikomatsakis** [comment]
@rustbot assign @djc 

I believe @djc was planning to write mentoring instructions here, so I'm assigning to them. We discussed it a bit on zulip; [my plan](https://zulip-archive.rust-lang.org/268952edition2021/55892TheWeeklyEditionoftheEditionUpdatesEdition5.html#235205451) and @djc's [somewhat simpler plan](https://zulip-archive.rust-lang.org/268952edition2021/55892TheWeeklyEditionoftheEditionUpdatesEdition5.html#235246192). Either of these seem great to me.

**Comment by djc** [comment]
The latest version of this is now in RFC 3114 ([rendered](https://github.com/djc/rfcs/blob/prelude-2021-traits/text/0000-prelude-2021-traits.md)).

**Comment by nikomatsakis** [comment]
Here are the instructions written in RFC 3114:

### Migration Lint

As for all edition changes, we will implement a migration lint to detect cases where code would break in the new edition. It includes a MachineApplicable suggestion for an alternative that will work in both the current and next edition.

The migration lint will be implemented as follows:

* Find method calls matching the name of one of the newly added traits' methods.
  This can be done either by hardcoding these method names or by setting up some
  kind of registry through the use of an attribute on the relevant traits.
* After method resolution, if:
  * The method matches one of the newly added methods' names, and
  * The originating trait is not from `core` or `std` (and/or does not
  match the originating trait), and
  * The originating trait is also implemented for the receiver's type,
    * Suggest a rewrite to disambiguating syntax (such as `foo.try_into()` to `TryInto::try_into(foo)`). If necessary, additional levels of (de)referencing
    might be needed to match the implementing type of the target trait.

Currently, diagnostics for trait methods that are not in scope suggest importing the originating trait. For traits that have become part of the prelude in a newer edition, the diagnostics should be updated such that they suggest upgrading to the latest edition as an alternative to importing the relevant trait.

**Comment by nikomatsakis** [comment]
Here are some mentoring instructions for how to do this.

To start, I would go with hard-coding the names of the methods for now. It's easier. Those methods are the methods from `TryFrom`, `TryInto`, and `FromIterator`. 

The method call resolution is is n the `rustc_typeck::check::method` module:

https://github.com/rust-lang/rust/blob/ef0ec303fa5ff6fa5654c7ea94322f3f9881f6c2/compiler/rustc_typeck/src/check/method/mod.rs#L170-L192

The call to [`lookup_probe`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/fn_ctxt/struct.FnCtxt.html#method.lookup_probe) is what selects which method is being invoked:

https://github.com/rust-lang/rust/blob/ef0ec303fa5ff6fa5654c7ea94322f3f9881f6c2/compiler/rustc_typeck/src/check/method/mod.rs#L198-L199

We want to check if the method name is in the desired set of methods. The method name can be found in `segment.ident`, [it is an `Ident`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/struct.PathSegment.html#structfield.ident). You can add the `try_from`, `try_into` idents etc in this table:

https://github.com/rust-lang/rust/blob/ef0ec303fa5ff6fa5654c7ea94322f3f9881f6c2/compiler/rustc_span/src/symbol.rs#L105-L123

similar to how we already have [`as_str`](https://github.com/rust-lang/rust/blob/ef0ec303fa5ff6fa5654c7ea94322f3f9881f6c2/compiler/rustc_span/src/symbol.rs#L284) etc.

You can then compare `segment.ident` against `sym::try_from` etc to see if the method has the desired name.

Assuming it does, we want to see if the method being invoked is from a trait in libstd/libcore. If it is, we don't have to give a warning, since we are already using the method we expect. Otherwise, we should. 

To determine what method is being invoked, we want to check the [`PickResult`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/method/probe/type.PickResult.html) which, if `Ok`, contains a [`Pick`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/method/probe/struct.Pick.html). This contains a field `kind` that has a [`PickKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/method/probe/enum.PickKind.html) -- `PickKind::Trait` corresponds to our case.

We then check the [`item`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/method/probe/struct.Pick.html#structfield.item) field to see what is invoked. If the [`item.def_id`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/assoc/struct.AssocItem.html#structfield.def_id) field comes from libstd or libcore, then no warning is required. You can determine that by checking the [`krate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/def_id/struct.DefId.html#structfield.krate) field of the `DefId`. Get the [`cstore`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.GlobalCtxt.html#structfield.cstore) from the tcx or something. (I'm not actually 100% sure the best way to do this step.)

**Comment by jam1garner** [comment]
Currently the big issue on the PR is that a majority of the breaking changes (by cases to handle, likely not by how common it is) look involve associated functions and thus something like this:

```rust
trait TryFromU8 {
    fn try_from(x: u8) -> Result<Self, !>;
}

impl TryFromU8 for u32 {
    fn try_from(x: u8) -> Result<Self, !> {
        x as u32
    }
}

let x = u32::try_from(3);
```

The part of typeck we were looking at only handles dot-call syntax and can only handle `try_into` (as it's the only method). When I get back to this I'm going to start looking for where associated functions are looked up for a given path during type checking, as that'd likely be a good place to target handling the associated function for the lint.

**Comment by nikomatsakis** [comment]
@jam1garner ah, good catch, I didn't consider this case. It should be relatively easy to manage this as well.

**Comment by m-ou-se** [comment]
Fixed by #85707 :tada:
